clustering/ee/cache/src/main/java/org/wildfly/clustering/ee/cache/ConcurrentManager.java

```diff
@@ -39,43 +40,47 @@
  */
 public class ConcurrentManager<K, V> implements Manager<K, V> {
 
-    private final Map<K, Map.Entry<Integer, V>> objects = new ConcurrentHashMap<>();
-    private final BiFunction<K, Map.Entry<Integer, V>, Map.Entry<Integer, V>> addFunction = new BiFunction<K, Map.Entry<Integer, V>, Map.Entry<Integer, V>>() {
+    private final Map<K, Map.Entry<Integer, VolatileReference<V>>> objects = new ConcurrentHashMap<>();
+    private final BiFunction<K, Map.Entry<Integer, VolatileReference<V>>, Map.Entry<Integer, VolatileReference<V>>> addFunction = new BiFunction<K, Map.Entry<Integer, VolatileReference<V>>, Map.Entry<Integer, VolatileReference<V>>>() {
         @Override
-        public Map.Entry<Integer, V> apply(K id, Map.Entry<Integer, V> entry) {
-            return entry != null ? new AbstractMap.SimpleImmutableEntry<>(Integer.valueOf(entry.getKey().intValue() + 1), entry.getValue()) : new VolatileEntry<>(new Integer(0));
+        public Map.Entry<Integer, VolatileReference<V>> apply(K id, Map.Entry<Integer, VolatileReference<V>> entry) {
+            Integer count = (entry != null) ? Integer.valueOf(entry.getKey().intValue() + 1) : new Integer(0);
```

Comment:
Perhaps s/new Integer(0)/Integer.valueOf(0)/g ?

Reply:
Good catch - given that this is the request thread, it's a worthwhile optimization.  I'll update this later.